aptos-move/framework/aptos-framework/sources/configs/stake.move

```diff
@@ -1013,34 +1009,37 @@ module aptos_framework::stake {
         pool_address: address,
         validator_set_config: &ValidatorSetConfiguration,
     ) acquires StakePool, StakePoolEvents, AptosCoinCapabilities, ValidatorConfig {
+        let stake_pool = borrow_global_mut<StakePool>(pool_address);
         let validator_config = borrow_global<ValidatorConfig>(pool_address);
         let num_missed_votes = *vector::borrow(&validator_perf.missed_votes, validator_config.validator_index);
         let num_blocks = validator_perf.num_blocks;
         let num_successful_votes = num_blocks - num_missed_votes;
 
-        let current_time = timestamp::now_seconds();
-        // Remaining lockup time needs to be at least one so the rewards calculation still works.
-        let remaining_lockup_time = 1;
-        let stake_pool = borrow_global_mut<StakePool>(pool_address);
-        if (current_time < stake_pool.locked_until_secs) {
-            remaining_lockup_time = stake_pool.locked_until_secs - current_time;
-        };
-
-        // Always merge rewards to active. If a validator's remaining lockup has already expired and they're withdrawing
-        // coins, their rewards would be 0 anyway as remaining lockup = 0.
         let rewards = mint_reward(
             validator.voting_power,
             num_blocks,
             num_successful_votes,
-            remaining_lockup_time,
             validator_set_config,
         );
         let rewards_amount = coin::value<AptosCoin>(&rewards);
-        coin::merge(&mut stake_pool.active, rewards);
+        // Merge rewards into active unless the validator is unlocking all of their stake.
+        if (coin::value(&stake_pool.active) > 0) {
+            coin::merge(&mut stake_pool.active, rewards);
+        } else {
+            coin::merge(&mut stake_pool.pending_inactive, rewards);
+        };
 
-        // Process any pending active or inactive stakes.
+        // Pending active stake can now be active.
         coin::merge<AptosCoin>(&mut stake_pool.active, coin::extract_all<AptosCoin>(&mut stake_pool.pending_active));
-        coin::merge<AptosCoin>(&mut stake_pool.inactive, coin::extract_all<AptosCoin>(&mut stake_pool.pending_inactive));
+
+        // Pending inactive stake is only fully unlocked and moved into inactive if the current lockup cycle has expired
+        let current_lockup_expiration = stake_pool.locked_until_secs;
+        if (timestamp::now_seconds() >= current_lockup_expiration) {
```

Comment:
This is a key part of the new flow - stakes can be unlocked anytime, but is still subject to the lockup and doesn't become withdrawable until the lockup has expired.

Reply:
I'll need to think about this more, it seems not very clean to mixed up on lockup time and epochs

Reply:
This does add some risk to the critical path (epoch transition) but does make the UX a lot better - delegations can easily be built on this for example, as they can request partial unlock and settle via a second call to withdraw when it's ready (remaining lockup time is also very easily obtainable via get_remaining_lockup_secs).